[content-visibility] Optimize lazy layout

rwlbuis opened the pull request:
#### 58d4460957a8b92c7109bddc65c5972a556aceb2
<pre>
[content-visibility] Optimize lazy layout
<a href="https://bugs.webkit.org/show_bug.cgi?id=263327">https://bugs.webkit.org/show_bug.cgi?id=263327</a>

Reviewed by Alan Baradlay.

Keep track of whether a hidden content-visibility subtree has been laid out using a dedicated bit
named everHadSkippedContentLayout, and only force layout if everHadSkippedContentLayout is false.
In order to make clearNeedsLayoutForSkippedContent work correctly the everHadSkippedContentLayout
bit is cleared for the whole content-visibility subtree.

This patch also merges isSkippedContentRootForLayout and isSkippedContentRoot since it was not
always clear which one to call and was a source of bugs. The new behaviour is now to always take layout into
account, but for size containment related code the static WebCore::isSkippedContentRoot is used
that does not take layout into account since size containment determination should be invariant to (forced) layout.

The test content-visibility-getboundingclientrect-on-children.html is added to verify layout is forced
only once for subtrees.

Canonical link: <a href="https://commits.webkit.org/274490@main">https://commits.webkit.org/274490@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b3eb956ff60fdd174456066120f11e74bf67e5d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35675 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14616 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38412 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32132 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36892 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11642 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30914 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36228 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12357 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31745 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10857 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10860 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39657 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32420 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32223 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36815 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11051 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8942 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34898 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12772 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31545 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8783 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11559 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11840 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on PerformanceTests/ContentVisibility/div-container-getboundingclientrect.html:
Would it be possible to add a corresponding WPT ? It would be great to test if this didn't timeout.

rwlbuis replied:
Sure, done :)

alanbaradlay reviewed (commented):
what's the logic behind repurposing the "ever had layout" bit? AFAICT this may affect some other, unrelated to c-v features. is that intentional?

rwlbuis committed:
[content-visibility] Optimize lazy layout
https://bugs.webkit.org/show_bug.cgi?id=263327

Reviewed by Alan Baradlay.

Keep track of whether a hidden content-visibility subtree has been laid out using a dedicated bit
named everHadSkippedContentLayout, and only force layout if everHadSkippedContentLayout is false.
In order to make clearNeedsLayoutForSkippedContent work correctly the everHadSkippedContentLayout
bit is cleared for the whole content-visibility subtree.

This patch also merges isSkippedContentRootForLayout and isSkippedContentRoot since it was not
always clear which one to call and was a source of bugs. The new behaviour is now to always take layout into
account, but for size containment related code the static WebCore::isSkippedContentRoot is used
that does not take layout into account since size containment determination should be invariant to (forced) layout.

The test content-visibility-getboundingclientrect-on-children.html is added to verify layout is forced
only once for subtrees.

Canonical link: https://commits.webkit.org/274490@main
